### PR TITLE
Optimise clone operation for incremental full cluster snapshots

### LINKED ISSUE #16295: [Feature Request] Optimise clone operation for incremental full cluster snapshots
### Is your feature request related to a problem? Please describe

During the development of shallow v1 snapshots, the clone operation code was modified to handle both shallow and existing full cluster incremental snapshots with the PR - https://github.com/opensearch-project/OpenSearch/pull/7496. However, in this PR, we have inadvertently introduced degraded experience for doc-rep cluster's clone snapshot operation where for each shard, there is dependency on repository data (index-N file for repository). If the index-N file size is above 512KB, then data needs to be downloaded from snapshot repository each time. This makes the clone snapshot operation slow. This can be reproduced by creating a lot of snapshots in a repo so that the index-N file size is above 512 KB.

### Describe the solution you'd like

For doc-rep clusters, we don't need to depend on the repository data to decide which type of snapshot was taken. For docrep clusters, we only support full cluster incremental snapshots. In this case, we will check whether the cluster is remote store enabled or not, and if not, then directly jump to the part of cloning without seeking the repository data for any shard.

### Related component

Storage:Snapshots

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ashking94**
### Description
In this PR, we have improved the logic for snapshot clone operations for full cluster incremental snapshots. The overall logic stays the same. There is no existing coverage for the snapshot clone operation. At this point, it is going to be huge effort to write unit tests for this flow. 

### Related Issues
Resolves #16295

### Check List
- ~[ ] Functionality includes testing.~
- ~[ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.~
- ~[ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.~

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1355]:
```suggestion
                if (!remoteStoreIndexShallowCopy) {
```

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java
Comment (dbwiddis) [inline server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1377]: I see this matches the original code, but I'm curious what happens to the ActionListener in the case this evaluates false. Is it ever completed?

**Review by dbwiddis — COMMENTED** [review]
Generally LGTM. One question.

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java
Comment (ashking94) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1355]: I have seen generally seen this trend of using == false in OpenSearch project and IMO this is probably more readable than !(boolean). Although I don't mind changing it as well.

Comment (dbwiddis) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1355]: It's definitely a style discussion and I wouldn't hold up approval for it.  The compiler will probably produce identical bytecode in either case.  You may see it a lot from code written by new(-ish) coders who have never been bit by accidentally using `=` instead of `==`.  The simpler code is, the less likely you are to make errors.

One reason I like the shorter form is it somewhat forces verbose and meaningful variable names (so I can read "if not a remote store index shallow copy" in this case).  Compare that to "if remote store index shallow copy is false" which is more "thinking "value of a variable in code" and less natural semantic logic and let you get away with poor variable naming.

It's generally frowned upon in style guides, e.g., [Sonar RSPEC-1125](https://rules.sonarsource.com/java/RSPEC-1125/) and [Checkstyle SimplifyBooleanExpression](https://checkstyle.sourceforge.io/checks/coding/simplifybooleanexpression.html).  Since I use a Sonar linter in my IDE, everytime I see these, they produce little blue squigglies and suggest I fix them. I tend to fix them when I come across them, but like to avoid avoid adding them into new code.

Again, not a big deal, just a common style preference.

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java
Comment (sachinpkale) [inline server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1356]: As I undersstand, except this if block, else is re-factoring. Is this correct?

**Review by sachinpkale — APPROVED** [review]

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java
Comment (ashking94) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1356]: That's right. No change in the code overall. We have just moved an if condition to earlier.

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java
Comment (ashking94) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1355]: thanks @dbwiddis, this does put a perspective. And as I think more, do you think if we should any of the static code analyser option widely used in industry. I have previously used find bugs and found it helpful to a large extent. Although this is a tangential discussion from this PR, but want to know your thoughts to start with and then can jump this idea across more maintainers through a GH issue. This way, we can try to fix all the similar issues.

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java
Comment (ashking94) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1377]: Firstly thanks for the thorough review, @dbwiddis. 

If the `currentlyCloning.add(repoShardId)` returns false, then it means that the underlying shard (with indexId "x" and shard no "y") is being cloned already. Also, runReadyClone method does not return any value to the caller function neither there is any interaction through the action listener. I am not fully sure, but it looks like we do not want to have more than 1 clone operation happening for the same shard. Since we have guard rails that prevent concurrent snapshot operations at cluster level, it probably appears to be defensive coding and not sure if we would ever hit the `currentlyCloning.add(repoShardId)` with value false. Let me know if you think otherwise from code side.

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1355]: @ashking94 It's a mixed bag.  I've considered proposing we add analysis using Sonar to the project, as it's been extremely helpful on my own open source project, and helps me find a lot of issues (potential NPE are particularly valuable).  And having the Sonar linter in my own IDE is helpful when writing/reviewing PRs.  But some of the recommendations are silly and useless, and some rely on recent JVMs but don't work well with older ones.   So it's hard to do a one-size-fits-all solution.

There's also the fact that we inherited a decades-old codebase and doing massive fixes of style preference isn't the best use of our time.  So I generally just focus on "new code" while an analyzer will find thousands of issues in the older code.

Probably worth a "best practices" documentation somewhere but we have enough issues with flakky tests etc. that putting an emphasis on minor style issues is probably not appropriate right now.

File: server/src/main/java/org/opensearch/snapshots/SnapshotsService.java
Comment (dbwiddis) [reply server/src/main/java/org/opensearch/snapshots/SnapshotsService.java:1377]: OK, looked at it again more closely.  Method naming bit me. `getCloneCompletionListener` actually creates a new listener, so those listeners are never created.  (So nit: should be named `create...()`) 😏 

So this LGTM.

**Review by dbwiddis — APPROVED** [review]

**Comment by dbwiddis** [comment]
One more question, does this need a CHANGELOG?

**Comment by ashking94** [comment]
> One more question, does this need a CHANGELOG?

Let me add this, the users may have stopped using clone or have never faced this issue due to lower volume of shards after this inadvertent bug. However, adding a change log may guide the user to explore and try this out.

**Review by dbwiddis — APPROVED** [review]

**Comment by ashking94** [comment]
Auto backport failed, will create a manual backport.

**Comment by ashking94** [comment]
Manual backport - https://github.com/opensearch-project/OpenSearch/pull/16303

**Pull request merged**
